### PR TITLE
Reduce learning rate on plateau

### PR DESCRIPTION
**Pull request by DanBmh**
There is a keras callback with this functionality which very often improves my accuracy (i mostly use it for image classification). So i thought this would be a nice feature to have here too.  First test runs were promising.

I had to change the way an early stop is triggered too, to ensure the training will not be stopped to fast.

### CODE REVIEW

File: util/flags.py [hunk 1]
```diff
@@ -135,9 +135,15 @@ def create_flags():
...
-    f.DEFINE_integer('es_steps', 4, 'number of validations to consider for early stopping. Loss is not stored in the checkpoint so when checkpoint is revived it starts the loss calculation from start at that point')
-    f.DEFINE_float('es_mean_th', 0.5, 'mean threshold for loss to determine the condition if early stopping is required')
-    f.DEFINE_float('es_std_th', 0.5, 'standard deviation threshold for loss to determine the condition if early stopping is required')
+    f.DEFINE_integer('es_epochs', 4, 'Number of epochs with no improvement after which training will be stopped. Loss is not stored in the checkpoint so when checkpoint is revived it starts the loss calculation from start at that point')
+    f.DEFINE_float('es_min_delta', 0.05, 'Minimum change in loss to qualify as an improvement.')
+
+    # Reduce learning rate on plateau
+
+    f.DEFINE_boolean('reduce_lr_on_plateau', True, 'Enable reducing the learning rate if a plateau is reached. This is the case if the validation loss did not improve for some epochs.')
+    f.DEFINE_integer('plateau_epochs', 2, 'Number of validations to consider for RLROP. Has to be smaller than the steps for early stopping')
```
Comment (reuben) [inline util/flags.py]: Reference the es_epochs flag by name here.

File: util/flags.py [hunk 2]
```diff
@@ -135,9 +135,15 @@ def create_flags():
...
+    f.DEFINE_float('plateau_reduction', 0.1, 'Factor to reduce the current learing rate if a plateau has occured.')
```
Comment (reuben) [inline util/flags.py]: nit: `Multiplicative factor to apply to the current learning rate if a plateau has occured.`

File: util/flags.py (hunk 1 above)
Comment (reuben) [reply util/flags.py]: In my experience training DeepSpeech models, `--plateau_epochs 2 --plateau_reduction 0.1` are not good defaults. I've seen the training vary for up to 15 epochs before continuing to decrease steadily. I would start with `--plateau_epochs 5 --plateau_reduction 0.9` or maybe `--plateau_epochs 10 --plateau_reduction 0.1`.

@kdavis-mozilla you've been training a lot of models lately, what do you think?

File: util/flags.py [hunk 3]
```diff
@@ -135,9 +135,15 @@ def create_flags():
     # Early Stopping
 
     f.DEFINE_boolean('early_stop', True, 'enable early stopping mechanism over validation dataset. If validation is not being run, early stopping is disabled.')
-    f.DEFINE_integer('es_steps', 4, 'number of validations to consider for early stopping. Loss is not stored in the checkpoint so when checkpoint is revived it starts the loss calculation from start at that point')
-    f.DEFINE_float('es_mean_th', 0.5, 'mean threshold for loss to determine the condition if early stopping is required')
-    f.DEFINE_float('es_std_th', 0.5, 'standard deviation threshold for loss to determine the condition if early stopping is required')
+    f.DEFINE_integer('es_epochs', 4, 'Number of epochs with no improvement after which training will be stopped. Loss is not stored in the checkpoint so when checkpoint is revived it starts the loss calculation from start at that point')
```
Comment (reuben) [inline util/flags.py]: How about just tying this number to `--plateau_epochs`? Just make it always be 2x that flag. Avoids configuration errors and it's a pretty good default I would say.

File: DeepSpeech.py [hunk 4]
```diff
@@ -468,7 +468,8 @@ def train():
     }
 
     # Building the graph
-    optimizer = create_optimizer()
+    learning_rate_tensor = tf.placeholder(tf.float32, shape=[])
```
[the remark this answers is not in the document]
Comment (DanBmh) [reply DeepSpeech.py]: Will this break the old checkpoints?

File: util/flags.py (hunk 3 above)
Comment (DanBmh) [reply util/flags.py]: I also thought about it. The downside is that you cannot customise it anymore. But i also dont know where this really makes a difference.

By the way, in keras early stopping and rlrop are two different callbacks.

File: DeepSpeech.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reuben) [reply DeepSpeech.py]: Yes.

[the remark this answers is not in the document]
Comment (DanBmh) [reply DeepSpeech.py]: Can we/you do this in another pull request then? Maybe shortly before you release a new model.

I need your released english checkpoint because im using it as base for my training in german.

[the remark this answers is not in the document]
Comment (reuben) [reply DeepSpeech.py]: We can also just hold this PR until shortly before we release a model, if you need that. One thing I'll note is that you could write some small amount of code to handle this case, by initializing the learning rate variable if it's not found in the checkpoint. But this will introduce conflicts with the transfer learning PR so I'd just wait.

[the remark this answers is not in the document]
Comment (lissyx) [reply DeepSpeech.py]: I have to admit i'd love to see that in 0.7.0, both because I'd want to experiment with it, and because that would give us more time, if it can't be landed in a compatible way, then let's wait for after 0.7/

[the remark this answers is not in the document]
Comment (reuben) [reply DeepSpeech.py]: We can fix the 0.7 checkpoint to have the variable in it as well, it should be quite simple.

[the remark this answers is not in the document]
Comment (DanBmh) [reply DeepSpeech.py]: Wont saving the learning rate in a variable make transfer learning more complicated as you have to modifiy the graph to change the learning rate back to a higher start value again?

[the remark this answers is not in the document]
Comment (reuben) [reply DeepSpeech.py]: You don't have to modify the graph, just assign a value to the variable. Transfer learning already requires extensive logic around checkpoint loading anyway so this would not be a complex feature.

File: util/flags.py (hunk 1 above)
Comment (DanBmh) [reply util/flags.py]: Im using this `--plateau_epochs 10 --plateau_reduction 0.1` now. This also the keras default.

File: DeepSpeech.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DanBmh) [reply DeepSpeech.py]: Saving the learning rate in a variable now

File: util/flags.py (hunk 3 above)
Comment (DanBmh) [reply util/flags.py]: I think a factor of 2 may be too small, because then the learning rate will be reduced only once before doing the early stop.

File: DeepSpeech.py [hunk 5]
```diff
@@ -253,11 +252,13 @@ def calculate_mean_edit_distance_and_loss(iterator, dropout, reuse):
 # we will use the Adam method for optimization (http://arxiv.org/abs/1412.6980),
 # because, generally, it requires less fine-tuning.
 def create_optimizer():
-    optimizer = tfv1.train.AdamOptimizer(learning_rate=FLAGS.learning_rate,
+    learning_rate_var = tf.Variable(FLAGS.learning_rate, trainable=False)
```
Comment (reuben) [inline DeepSpeech.py]: This should be using `tfv1.get_variable('learning_rate', initializer=FLAGS.learning_rate, trainable=False)`

File: DeepSpeech.py [hunk 6]
```diff
@@ -652,29 +655,40 @@ def __call__(self, progress, data, **kwargs):
...
+                        log_info('Early stop triggered as the loss did not improve the last {} epochs'.format(
+                            epochs_without_improvement))
+                        break
+
+                    # Reduce learning rate on plateau
+                    if (FLAGS.reduce_lr_on_plateau and
+                            epochs_without_improvement % FLAGS.plateau_epochs == 0 and epochs_without_improvement > 0):
+                        # If the learning rate was reduced and there is still no improvement
+                        # wait FLAGS.plateau_epochs before the learning rate is reduced again
+
```
Comment (reuben) [inline DeepSpeech.py]: nit: remove this empty line

File: DeepSpeech.py (hunk 5 above)
Comment (reuben) [reply DeepSpeech.py]: Also, you can create this variable in the caller and pass it to `create_optimizer`, it's cleaner than returning multiple values here.

**Comment by reuben** [comment]
I'm going to make sure this passes tests and deal with the missing variable stuff in a separate PR.
